File: tests/go/ice-test/flow/score_flow.go
// Package flow provides flow leaf nodes for testing.
package flow

import (
	"context"

	icecontext "github.com/zjn-zjn/ice/sdks/go/context"
)

// ScoreFlow checks if roam[key] >= score.
type ScoreFlow struct {
	Score float64 `json:"score" ice:"name:分数阈值,desc:判断分数的阈值"`
	Key   string  `json:"key" ice:"name:取值键,desc:从roam中取值的键名"`
}

// DoFlow implements the LeafFlow interface.
func (s *ScoreFlow) DoFlow(ctx context.Context, roam *icecontext.Roam) bool {
	value := roam.ValueDeep(s.Key).Float64Or(0)
	return value >= s.Score
}

// ScoreFlow2 is another variant of score checking.
type ScoreFlow2 struct {
	Score float64 `json:"score" ice:"name:分数阈值,desc:判断分数的阈值"`
	Key   string  `json:"key" ice:"name:取值键,desc:从roam中取值的键名"`
}

// DoFlow implements the LeafFlow interface.
func (s *ScoreFlow2) DoFlow(ctx context.Context, roam *icecontext.Roam) bool {
	value := roam.ValueDeep(s.Key).Float64Or(0)
	return value >= s.Score
}

File: tests/python/ice-test/flow/score_flow.py
"""Score flow leaf nodes for testing."""

from typing import Annotated

import ice
from ice import IceField, Roam


@ice.leaf(
    "com.ice.test.flow.ScoreFlow",
    name="分数判断节点",
    desc="取出roam中的值比较大小",
    alias=["score_flow"],
)
class ScoreFlow:
    """Checks if roam[key] >= score."""
    score: Annotated[float, IceField(name="分数阈值", desc="判断分数的阈值")] = 0.0
    key: Annotated[str, IceField(name="取值键", desc="从roam中取值的键名")] = ""
    
    def do_roam_flow(self, roam: Roam) -> bool:
        value = roam.get_multi(self.key)
        if value is None:
            return False
        
        try:
            value_score = float(value)
        except (ValueError, TypeError):
            return False
        
        return value_score >= self.score


@ice.leaf(
    "com.ice.test.flow.ScoreFlow2",
    name="分数判断节点2",
    desc="另一个分数判断",
    alias=["score_flow_2"],
)
class ScoreFlow2:
    """Another variant of score checking."""
    score: Annotated[float, IceField(name="分数阈值", desc="判断分数的阈值")] = 0.0
    key: Annotated[str, IceField(name="取值键", desc="从roam中取值的键名")] = ""
    
    def do_roam_flow(self, roam: Roam) -> bool:
        value = roam.get_multi(self.key)
        if value is None:
            return False
        
        try:
            value_score = float(value)
        except (ValueError, TypeError):
            return False
        
        return value_score >= self.score
